src/gtc/passes/oir_optimizations/inlining.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import copy as cp
+from typing import Any, Dict, Set
+
+from eve import NodeTranslator, NodeVisitor
+from eve.concepts import NOTHING
+from gtc import oir
+
+
+class MaskCollector(NodeVisitor):
+    """Collects the boolean expressions defining mask statements that are boolean fields."""
+
+    def visit_AssignStmt(
+        self,
+        node: oir.AssignStmt,
+        *,
+        masks_to_inline: Dict[str, oir.Expr],
+    ) -> None:
+        if node.left.name in masks_to_inline:
+            assert masks_to_inline[node.left.name] is None
+            masks_to_inline[node.left.name] = node.right
+
+    def visit_MaskStmt(
+        self, node: oir.MaskStmt, *, masks_to_inline: Dict[str, oir.Expr], **kwargs: Any
+    ) -> None:
+        if isinstance(node.mask, oir.FieldAccess) and node.mask.name in masks_to_inline:
+            # Find all reads in condition
+            condition_reads: Set[str] = (
+                masks_to_inline[node.mask.name]
+                .iter_tree()
+                .if_isinstance(oir.FieldAccess, oir.ScalarAccess)
+                .getattr("name")
+                .to_set()
+            )
+            # Find all writes in body
+            body_writes: Set[str] = set()
+            body_writes |= set(
+                [child.left.name for child in node.body if isinstance(child, oir.AssignStmt)]
+            )
```

Comment:
Warning: I have not actually tested this suggestion.

```suggestion
            body_writes = node.iter_tree().if_isinstance(oir.AssignStmt).getattr("left").getattr("name").to_set()
```